tests/templates/test_state_preparations/test_arbitrary_state_prep.py

```diff
@@ -0,0 +1,351 @@
+
+        weights = np.zeros(14)
+        weights[13] = np.pi / 2
+
+        @qml.qnode(qubit_device_3_wires)
+        def circuit(weights):
+            qml.templates.ArbitraryStatePreparation(weights, [0, 1, 2])
+            return qml.expval(qml.PauliZ(0))
+
+        circuit(weights)
+
+        assert np.allclose(circuit.device.state, GHZ_state, atol=tol, rtol=0)
+
+    def test_even_superposition_generation(self, qubit_device_3_wires, tol):
+        """Test that the template prepares an even superposition state."""
+        even_superposition_state = np.ones(8) / np.sqrt(8)
+
+        weights = np.zeros(14)
+        weights[1] = np.pi / 2
+        weights[3] = np.pi / 2
+        weights[5] = np.pi / 2
+
+        @qml.qnode(qubit_device_3_wires)
+        def circuit(weights):
+            qml.templates.ArbitraryStatePreparation(weights, [0, 1, 2])
+
+            return qml.expval(qml.PauliZ(0))
+
+        circuit(weights)
+
+        assert np.allclose(circuit.device.state, even_superposition_state, atol=tol, rtol=0)
+
+    def test_custom_wire_labels(self, tol):
+        """Test that template can deal with non-numeric, nonconsecutive wire labels."""
+        weights = np.random.random(size=(2 ** 4 - 2))
+
+        dev = qml.device("default.qubit", wires=3)
+        dev2 = qml.device("default.qubit", wires=["z", "a", "k"])
+
+        @qml.qnode(dev)
+        def circuit():
+            qml.templates.ArbitraryStatePreparation(weights, wires=range(3))
+            return qml.expval(qml.Identity(0))
+
+        @qml.qnode(dev2)
+        def circuit2():
+            qml.templates.ArbitraryStatePreparation(weights, wires=["z", "a", "k"])
+            return qml.expval(qml.Identity("z"))
+
+        circuit()
+        circuit2()
+
+        assert np.allclose(dev.state, dev2.state, atol=tol, rtol=0)
+
+
+class TestInputs:
+    """Test inputs and pre-processing."""
+
+    def test_exception_wrong_dim(self):
+        """Verifies that exception is raised if the
+        number of dimensions of features is incorrect."""
+        dev = qml.device("default.qubit", wires=3)
+
+        @qml.qnode(dev)
+        def circuit(weights):
+            qml.templates.ArbitraryStatePreparation(weights, wires=range(3))
+            return qml.expval(qml.PauliZ(0))
+
+        with pytest.raises(ValueError, match="Weights tensor must be of shape"):
+            weights = np.zeros(12)
+            circuit(weights)
+
+
+class TestAttributes:
+    """Tests additional methods and attributes"""
+
+    @pytest.mark.parametrize(
+        "n_wires, expected_shape",
+        [
+            (3, (14,)),
+            (1, (2,)),
+            (2, (6,)),
+        ],
+    )
+    def test_shape(self, n_wires, expected_shape):
+        """Test that the shape method returns the correct shape of the weights tensor"""
+
+        shape = qml.templates.ArbitraryStatePreparation.shape(n_wires)
+        assert shape == expected_shape
+
+
+def circuit_template(weights):
```

Comment:
Should these be fixtures? Just curious.

Reply:
I think it's ok if they are just functions defined somewhere...since they change in every test, so it would be no use to make them global fixtures...Or did I understand wrong?

I am never sure if I should put them into the test class, but I find it so ugly to refer to them with `self`...